src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+
+use anyhow::anyhow;
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::complete_multipart_upload::CompleteMultipartUploadError;
+use aws_sdk_s3::operation::create_multipart_upload::CreateMultipartUploadError;
+use aws_sdk_s3::operation::upload_part::UploadPartError;
+use aws_sdk_s3::primitives::ByteStream;
+use aws_sdk_s3::types::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::Client;
+use aws_types::sdk_config::SdkConfig;
+use bytes::Bytes;
+use bytesize::ByteSize;
+use http::Uri;
+use mz_ore::error::ErrorExt;
+
+/// A multi part uploader which can upload multiple parts
+/// and keeps track of state to eventually finish the upload process.
+/// The caller does not need to know the final number of parts beforehand.
+///
+/// The caller should get an instance by calling `S3MultiPartUploader::new` first.
+/// Then start the multi-part upload by calling `start`.
+/// Each part can be added by calling `upload_part`, and can be called one or more times
+/// and eventually finish the multi part upload by calling `finish` method.
+pub struct S3MultiPartUploader<T = Initialized> {
+    client: Client,
+    bucket: String,
+    path: String,
+    // State to track of ongoing multi part upload.
+    state: T,
+}
+/// The default state just after an `S3MultiPartUploader` is created.
+pub struct Initialized;
+/// State after a multi part upload is in progress.
+pub struct InProgress {
+    upload_id: String,
+    // Mutable state to track the in progress multi part upload.
+    etags: BTreeMap<i32, String>,
+    current_part_number: i32,
+    total_bytes_uploaded: u64,
+}
+
+/// The smallest allowable part number (inclusive).
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const MIN_PART_NUM: i32 = 1;
+/// The largest allowable part number (inclusive).
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const MAX_PART_NUM: i32 = 10_000;
+/// The minimum size of a part in a multipart upload.
+///
+/// This minimum doesn't apply to the last chunk, which can be any size.
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const MIN_PART_SIZE: ByteSize = ByteSize::mb(5);
+
+pub struct MultiPartProgress {
```

Comment:
nit: recommend calling this a "completed upload" or somesuch, since AFAICT this reflects a completed upload, not partial progress of an upload.

```suggestion
pub struct CompletedUpload {
```

Reply:
updated